feat: lark parser

camriddell opened the pull request:
Uses [lark](https://github.com/lark-parser/lark) to construct a grammar and parser for the query strings.

- Adds a dependency on lark
- Adds a grammar to parse the current query features
- defines entry points in `search.py`
    - `def parse` → to return a graph composed of `parse.BinaryOp`s and `SearchNode`s
    - `def search` → convenience to `parse` a query, convert the result to a `narwhals.Expr` and use it to filter passed in data

To Do
- [x] Add proper exceptions
- [x] Write tests that are actually tests
    - [x] thoroughly test grammar features
- [ ] add comments to grammar file?
    - [ ] create human readable version of grammar for documentation
- [ ] implement fallback comparison if no terms are found in a given query string (this is not supported at the grammar level as of now)

camriddell committed:
WIP

camriddell committed:
uv add lark

williambdean commented on src/frame_search/parse.py:
```diff
@@ -0,0 +1,153 @@
+VALUE_DEPENDENT_COMPARATORS = defaultdict(
+    lambda: operator.eq,
+    {
+        str: lambda col, value: col.str.contains(value),
```
I had this using to_lowercase and lower on the string value just to make it easier while using

# Column
["Bob", "Bobby", "Billy Bob"]
search("name:bob") # Would work for all

williambdean commented on src/frame_search/parse.py:
```diff
@@ -0,0 +1,153 @@
+    def BACKTICKED_KEY(self, token: Token) -> nw.Expr:
+        return Column(token.value[1:-1])
+
+    def BARE_KEY(self, token: Token) -> nw.Expr:
+        return Column(token.value)
```
The previous behavior treats these not as Column but as a value to match in a *default* column. For example, having a default column of `state` allows users to just use `search("California")` for matches. Those default and columns aliases might be needed in an init of this Transformer

williambdean replied:
This might be wrong spot to point this out though

williambdean reviewed (commented):
Just some comments on how the previous functionality worked

camriddell committed:
add pyright to dev for type checking

camriddell committed:
feat: improve Lark grammar specification

The new grammar introduces standard comparators and
allows the colon comparator to use value-specific comparisons.
The rules are as follows:
- The left side of any comparator is always parsed as a column name
- The right side of any comparator can be a value, or composition of
  values
- numeric comparisons can be performed via standard comparators <>=!
- the ":" comparator is value-depdendent
        - "x,y,z" (isin) supports checking multiple values
        - 2..6 (range) supports numeric/date range comparisons
        - "abe" (escaped strings) are lowercase matched

camriddell committed:
Merge remote-tracking branch 'upstream/main' into feat/lark-parser

camriddell committed:
remove unused import

camriddell committed:
update README for grammar explanation

camriddell committed:
add comment for operand flipping

camriddell committed:
replace typing.Self with TypeVar

camriddell committed:
typing syntax to python3.9

camriddell commented:
Made some updates to flesh out the grammar a bit more. There are two major features that still need to be integrated

1. default_column: allow queries that consist of a single value to be evaluated against a default column.
2. column_mapping: allow modification of column names to make querying via a string more convenient.

Let me know if there is anything else as well that you can find/think of!

williambdean commented:
Thanks @camriddell! Your work here was the foundation for #43. Your original Lark grammar commit landed in `main` under your name (360f06b), and you're credited as co-author on the boolean operators commit that built on top of it. Closing this in favour of the merged PR.
